re.c

```diff
@@ -2348,6 +2348,82 @@ match_named_captures(VALUE match)
     return hash;
 }
 
+/*
+ *  call-seq:
+ *    deconstruct_keys(array_of_names) -> hash
+ *
+ *  Returns a hash of the named captures for the given names.
+ *
+ *    m = /(?<hours>\d{2}):(?<minutes>\d{2}):(?<seconds>\d{2})/.match("18:37:22")
+ *    m.deconstruct_keys([:hours, :minutes]) # => {:hours => "18", :minutes => "37"}
+ *    m.deconstruct_keys(nil) # => {:hours => "18", :minutes => "37", :seconds => "22"}
+ *
+ *  Returns an empty hash of no named captures were defined:
+ *
+ *    m = /(\d{2}):(\d{2}):(\d{2})/.match("18:37:22")
+ *    m.deconstruct_keys(nil) # => {}
+ *
+ */
+static VALUE
+match_deconstruct_keys(VALUE match, VALUE keys)
+{
+    VALUE h, names;
+    long i;
+
+    match_check(match);
+
+    if (NIL_P(RMATCH(match)->regexp)) {
+        return rb_hash_new_with_size(0);
+    }
+
+    names = rb_reg_names(RMATCH(match)->regexp);
```

Comment:
It is better to delay calling `rb_reg_names` until it is really needed.
Pseudo code is here:

```
if (NIL_P(keys)) {
  VALUE names = rb_reg_names(RMATCH(match)->regexp);
  ...
}

...

if (onig_number_of_names(RREGEXP_PTR(RMATCH(match)->regexp))) < RARRAY_LEN(keys)) {
  return rb_hash_new_with_size(0);
}
```

(In addition to that, I think `rb_reg_names` in `if (NIL_P(keys)) {` could be replaced by `onig_foreach_name` so that we can avoid `rb_ary_new_capa` call in `rb_reg_names`.)